packages/core/fs/src/MemoryFS.js

```diff
@@ -0,0 +1,378 @@
+      if (this.dirs.has(dir)) {
+        break;
+      }
+
+      this.dirs.add(dir);
+      dir = path.dirname(dir);
+    }
+  }
+
+  async rimraf(filePath: FilePath) {
+    filePath = this._normalizePath(filePath);
+
+    if (this.dirs.has(filePath)) {
+      let dir = filePath + path.sep;
+      for (let filePath of this.files.keys()) {
+        if (filePath.startsWith(dir)) {
+          this.files.delete(filePath);
+        }
+      }
+
+      for (let dirPath of this.dirs.keys()) {
+        if (dirPath.startsWith(dir)) {
+          this.dirs.delete(dirPath);
+        }
+      }
+    }
+
+    this.dirs.delete(filePath);
+    this.files.delete(filePath);
+  }
+
+  async ncp(source: FilePath, destination: FilePath) {
+    source = this._normalizePath(source);
+
+    if (this.dirs.has(source)) {
+      this.dirs.add(destination);
+
+      let dir = source + path.sep;
+      for (let dirPath of this.dirs) {
+        if (dirPath.startsWith(dir)) {
+          let destName = path.join(destination, dirPath.slice(dir.length));
+          this.dirs.add(destName);
+        }
+      }
+
+      for (let [filePath, buffer] of this.files) {
+        if (filePath.startsWith(dir)) {
+          let destName = path.join(destination, filePath.slice(dir.length));
+          this.files.set(destName, buffer);
+        }
+      }
+    } else {
+      await this.copyFile(source, destination);
+    }
+  }
+
+  createReadStream(filePath: FilePath) {
+    let fs = this;
+    class ReadStream extends Readable {
+      _read() {
+        fs.readFile(filePath).then(
+          res => {
+            this.push(res);
+            this.push(null);
+          },
+          err => {
+            this.emit('error', err);
+          }
+        );
+      }
+    }
+
+    return new ReadStream();
+  }
+
+  createWriteStream(filePath: FilePath) {
+    let fs = this;
+    let buffer = Buffer.alloc(0);
+    class WriteStream extends Writable {
+      _write(
+        chunk: Buffer | string,
+        encoding: any,
+        callback: (error?: Error) => void
+      ) {
+        let c = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding);
+        buffer = Buffer.concat([buffer, c]);
+        callback();
+      }
+
+      _final(callback: (error?: Error) => void) {
+        fs.writeFile(filePath, buffer).then(callback);
+      }
+    }
+
+    return new WriteStream();
+  }
+
+  async realpath(filePath: FilePath) {
+    filePath = this._normalizePath(filePath);
+    return filePath;
+  }
+
+  async exists(filePath: FilePath) {
+    filePath = this._normalizePath(filePath);
+    return this.files.has(filePath) || this.dirs.has(filePath);
+  }
+}
+
+function createStat(size, isDir) {
+  return {
+    dev: 0,
+    ino: 0,
+    mode: 0,
+    nlink: 0,
+    uid: 0,
+    gid: 0,
+    rdev: 0,
+    size,
+    blksize: 0,
+    blocks: 0,
+    atimeMs: 0,
+    mtimeMs: 0,
+    ctimeMs: 0,
+    birthtimeMs: 0,
+    atime: new Date(0),
+    mtime: new Date(0),
+    ctime: new Date(0),
```

Comment:
How does the cache behave in this case (with MemoryFS as input fs)? The input files wouldn't have any change time info.

Reply:
Yeah none of that info was needed by Parcel itself, so I didn't implement it for now. Could be added later though.